File: common/src/main/java/cn/breadnicecat/candycraftce/block/blockentity/CBlockEntityBuilder.java
package cn.breadnicecat.candycraftce.block.blockentity;

import cn.breadnicecat.candycraftce.CandyCraftCE;
import com.mojang.datafixers.types.Type;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier;
import net.minecraft.world.level.block.entity.BlockEntityType.Builder;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.Objects;
import java.util.function.Supplier;

import static cn.breadnicecat.candycraftce.utils.ResourceUtils.prefix;

/**
 * Created in 2024/1/30 23:23
 * Project: candycraftce
 *
 * @author <a href="https://github.com/Bread-Nicecat">Bread_NiceCat</a>
 * <p>
 */
public class CBlockEntityBuilder<B extends BlockEntity> {
	public final String name;
	private final BlockEntitySupplier<B> factory;
	private Supplier<? extends Block>[] valid;
	@Nullable
	private Type<?> dsl;

	private CBlockEntityBuilder(String name, BlockEntitySupplier<B> factory) {
		this.name = name;
		this.factory = factory;
	}

	public static <B extends BlockEntity> CBlockEntityBuilder<B> create(String name, BlockEntitySupplier<B> factory) {
		return new CBlockEntityBuilder<>(name, factory);
	}

	@SafeVarargs
	public final CBlockEntityBuilder<B> setValidBlocks(Supplier<? extends Block>... blocks) {
		this.valid = blocks;
		return this;
	}

	public CBlockEntityBuilder<B> setDSL(Type<?> dsl) {
		this.dsl = dsl;
		return this;
	}

	public BlockEntityEntry<B> save() {
		Objects.requireNonNull(valid, "No valid blocks");
		return register(name, () -> Builder.of(factory,
				Arrays.stream(valid).map(Supplier::get).toArray(Block[]::new)
		).build(dsl));
	}

	private static <B extends BlockEntity> BlockEntityEntry<B> register(String name, Supplier<BlockEntityType<B>> factory) {
		return new BlockEntityEntry<>(CandyCraftCE.register(BuiltInRegistries.BLOCK_ENTITY_TYPE, prefix(name), factory));
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/block/CBlockEntityBuilder.kt
package cn.breadnicecat.candycraftce.core.block

import cn.breadnicecat.candycraftce.utils.CLogUtils
import cn.breadnicecat.candycraftce.utils.CUtils.modLoc
import cn.breadnicecat.candycraftce.utils.CUtils.register
import com.mojang.datafixers.types.Type
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier

/**
 * Created in 2024/1/30 23:23
 * Project: candycraftce
 * 
 * @author [Bread_NiceCat](https://github.com/Bread-Nicecat)
 * 
 * 
 */
class CBlockEntityBuilder<B : BlockEntity> private constructor(
    val name: String,
    private val factory: BlockEntitySupplier<B>,
) {
    private lateinit var valid: Array<out Block>
    private var dsl: Type<*>? = null

    @SafeVarargs
    fun setValidBlocks(vararg blocks: Block): CBlockEntityBuilder<B> {
        this.valid = blocks
        return this
    }

    fun setDSL(dsl: Type<*>): CBlockEntityBuilder<B> {
        this.dsl = dsl
        return this
    }

    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")

    fun save(): BlockEntityType<B> {
        val id = name.modLoc()
        CLogUtils.logRegister("BlockEntity", id)
        return BuiltInRegistries.BLOCK_ENTITY_TYPE.register(
            id,
            BlockEntityType.Builder.of(factory, *valid).build(dsl)
        )

    }

    companion object {
        fun <B : BlockEntity> create(name: String, factory: BlockEntitySupplier<B>): CBlockEntityBuilder<B> {
            return CBlockEntityBuilder(name, factory)
        }
    }
}
